### DISCUSSION TITLE
How to use MAISI's VAE encoder to encode a CT?

### DISCUSSION THREAD

**Discussion by paproaskwar** [Project-MONAI/tutorials#1985] [Q&A]
Really appreciate your wonderful job!!!

Here is my code when I use AutoencoderKlMaisi to encode a CT:

The code looks good, but when I save the latent(z_mu) as a .nii file and displayed in 3D Slicer, the latent looks pretty noisy.
is that normal?

    autoencoder = define_instance(args, "autoencoder_def").to(device)
    checkpoint_autoencoder = torch.load(args.trained_autoencoder_path, weights_only=True)
    autoencoder.load_state_dict(checkpoint_autoencoder)
    autoencoder.eval()

    with torch.no_grad(), torch.cuda.amp.autocast():
         z_mu, z_sigma = autoencoder.encode(images)
![捕获](https://github.com/user-attachments/assets/a31893f6-57a4-4635-ab57-9840cb350d13)

**Comment by Can-Zhao** [comment]
Yes

**Reply by sara-create** [reply]
By the way I have a question  , the embeddings of the labels are learned directly with the training script of the ControlNet ? Or there is a script for its training that I don't see.

**Reply by Can-Zhao** [reply]
There is no learned embeddings of the labels used for ControlNet. You can find how labels are processed in controlnet training script https://github.com/Project-MONAI/tutorials/blob/8b90a16d82a463551d81dd4da200f3d406c6843e/generation/maisi/scripts/train_controlnet.py#L196

**Reply by sara-create** [reply]
Thank you for your response. If I am not mistaken the labels are also sent into the latent space . In the ControlNet definition ( monai->networks->nets->controlnet.py ) there is a block called ( ControlNetConditioningEmbedding : Network to encode the conditioning into a latent space ) . My question is how is this block trained . Is it trained separately from the ControlNet or optimized with it. In the forward of the controlnet we call it through : controlnet_cond = self.controlnet_cond_embedding(controlnet_cond).

**Reply by Can-Zhao** [reply]
It's inside the controlnet, so it was trained together with the controlnet, you won't need to worry about it
